lib/adhearsion/events.rb

```diff
@@ -79,10 +78,24 @@ def refresh!
       end
 
       def clear
-        @queue = nil
+        kill!
+        @_executor = nil
         Handler.instance.clear_handlers
       end
+
+      def kill!
+        @_executor.kill if @_executor
+      end
+
+      def stop!
+        @_executor.shutdown if @_executor
+      end
+
     end
 
   end
 end
+
+Adhearsion::Events.register_callback :shutdown do
+  Adhearsion::Events.kill!
+end
```

Comment:
The minor cosmetic issue we've seen (which you were the first to note) is that the shutdown of event system is not graceful and triggers an ERROR upon shutdown when using Asterisk.  This is superficial and doesn't have any adverse behavior, at least none that I've seen.

But nevertheless, in my partial attempt in dissecting what's going on I think what happens is:
1. The `:shutdown` event triggers the Adhearsion::Events system to kill! itself.
2. The `:shutdown` event also triggers the Asterisk Rayo client to Celluloid#terminate its Asterisk Translator and ami_client.  This somehow leads ami_client to trigger a `Celluloid: CLEANUP CRASHED!`, which triggers the Adhearsion::Events `:exception` event.
3. For better or for worse (personally I think it's better as it is 👍), `Events#executor` is resilient enough to revive itself if new events come in, and does so.  (I think this is good, as we don't lose exception messages this way).
4. For some unfortunate reason however, when `Events#trigger` calls `Concurrent::ThreadPoolExecutor#post` it fails with `Concurrent::RejectedExecutionError`.

```
[2019-07-20 20:41:00.050] DEBUG Celluloid: Terminating task: type=:call, meta={:method_name=>:send_track_calls_to_actors!}, status=:sleeping
Celluloid::TaskFiber backtrace unavailable. Please try `Celluloid.task_class = Celluloid::TaskThread` if you need backtraces here.
[2019-07-20 20:41:00.060] DEBUG Celluloid: Terminating task: type=:call, meta={:method_name=>:run}, status=:iowait
Celluloid::TaskFiber backtrace unavailable. Please try `Celluloid.task_class = Celluloid::TaskThread` if you need backtraces here.
[2019-07-20 20:41:00.069] ERROR Celluloid: CLEANUP CRASHED!
IOError: closed stream
org/nio4r/Nio4r.java:188:in `deregister'
org/nio4r/Nio4r.java:462:in `close'
org/nio4r/Nio4r.java:453:in `close'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/celluloid-io-0.16.2/lib/celluloid/io/reactor.rb:52:in `wait'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/celluloid-io-0.16.2/lib/celluloid/io/reactor.rb:21:in `wait_readable'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/celluloid-io-0.16.2/lib/celluloid/io.rb:53:in `wait_readable'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/celluloid-io-0.16.2/lib/celluloid/io/stream.rb:30:in `wait_readable'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/celluloid-io-0.16.2/lib/celluloid/io/stream.rb:43:in `block in sysread'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/celluloid-io-0.16.2/lib/celluloid/io/stream.rb:389:in `synchronize'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/celluloid-io-0.16.2/lib/celluloid/io/stream.rb:39:in `sysread'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/celluloid-io-0.16.2/lib/celluloid/io/stream.rb:128:in `readpartial'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/ruby_ami-2.4.0/lib/ruby_ami/stream.rb:43:in `block in run'
org/jruby/RubyKernel.java:1425:in `loop'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/ruby_ami-2.4.0/lib/ruby_ami/stream.rb:43:in `run'
org/jruby/RubyKernel.java:1953:in `public_send'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/calls.rb:26:in `dispatch'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/calls.rb:137:in `dispatch'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/cell.rb:60:in `block in invoke'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/cell.rb:71:in `block in task'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/actor.rb:357:in `block in task'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/tasks.rb:57:in `block in initialize'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/tasks/task_fiber.rb:14:in `block in create'
[2019-07-20 20:41:00.078] ERROR Celluloid: EXCEPTION HANDLER CRASHED:
Concurrent::RejectedExecutionError: Concurrent::RejectedExecutionError
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/abstract_executor_service.rb:76:in `handle_fallback'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/concurrent-ruby-1.1.5/lib/concurrent/executor/java_executor_service.rb:31:in `post'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/adhearsion-250a4e6beb47/lib/adhearsion/events.rb:51:in `trigger'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/adhearsion-250a4e6beb47/lib/adhearsion.rb:116:in `block in <main>'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/logger.rb:66:in `block in crash'
org/jruby/RubyArray.java:1792:in `each'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/logger.rb:64:in `crash'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/actor.rb:348:in `cleanup'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/actor.rb:329:in `shutdown'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/actor.rb:164:in `run'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/actor.rb:130:in `block in start'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/thread_handle.rb:13:in `block in initialize'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/actor_system.rb:32:in `block in get_thread'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/bundler/gems/celluloid-df512cbbe68d/lib/celluloid/internal_pool.rb:132:in `block in create'
/srv/phone/apptastic/shared/bundle/jruby/2.5.0/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
[2019-07-20 20:41:00.095] DEBUG Celluloid: Terminating task: type=:call, meta={:method_name=>:run}, status=:iowait
Celluloid::TaskFiber backtrace unavailable. Please try `Celluloid.task_class = Celluloid::TaskThread` if you need backtraces here.
```

Reply:
yeah, its ugly but as you pointed out - we do not loose information and know to expect the `RejectedExecutionError`. there might be a better cleaner (without log noise) way, but it would require some experimentation and whether there isn't a possibility of waiting forever. in which case there needs to be a graceful timeout ... believe there isn't enough of the native APIs exposed with `Concurrent` to do that without resorting to (JRuby specific) internals.

Reply:
... or maybe there is: https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent/executor/java_executor_service.rb#L34